nncf/common/tensor_statistics/collectors.py

```diff
@@ -0,0 +1,226 @@
+"""
+ Copyright (c) 2021 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from abc import ABC
+from abc import abstractmethod
+from collections import deque
+from typing import Tuple
+
+from nncf.common.tensor import NNCFTensor
+from nncf.common.tensor_statistics.statistics import MinMaxTensorStatistic
+
+ReductionShape = Tuple[int]
+
+
+class TensorStatisticCollectorBase(ABC):
+    """Collector estimate statistics at the quantization point based on the provided reduction shape."""
+
+    def __init__(self, reduction_shape: ReductionShape = None, num_samples: int = None):
+        self._reduction_shape = reduction_shape
+        self._enabled = True
+        self._collected_samples = 0
+        self._num_samples = num_samples
+
+    def register_input(self, x):
```

Comment:
Type hints

Reply:
Input here is either `torch.Tensor` or `tf.Tensor`. Conversion to FW-specific NNCFTensor is done within FW-redefined `_register_input()`. Do you propose to change it?

Reply:
In this cases we use `TypeVar` from `typing` as a stand-in for template type placeholder. Take a look at other uses of `TypeVar` throughout NNCF code and try to apply it here.

Reply:
Fixed, thanks!